EIPS/eip-5450.md

```diff
@@ -0,0 +1,145 @@
+
+In particular this extended code validation eliminates the need for EVM stack underflow checks done for every executed instruction. It also prevents deploying the code that can be statically proven to require more than 1024 stack items.
+
+## Specification
+
+### Code validation
+
+*Remark:* We rely on the notions of *data stack* and *type section* as defined by [EIP-4750](./eip-4750.md).
+
+Code section validation rules as defined by [EIP-3670](./eip-3670.md) (which has been extended by [EIP-4200](./eip-4200.md) and [EIP-4750](./eip-4750.md)) are extended again to include the instruction flow traversal procedure, where every possible code path is examined, while keeping track of minimum required data stack height and verifying the following properties:
+
+1. For each reachable instruction in the section, data stack height when executing it is the same for all possible code paths going through this instruction.
+2. For each instruction, data stack always has enough items, i.e. stack underflow is invalid.
+3. For `CALLF` instruction data stack has enough items to use as input arguments to a called function according to its type defined in the type section.
+4. For every terminating instruction except `RETF`, data stack is empty after executing it.
+5. For `RETF` instruction data stack before executing it has exactly `n` items to use as output values returned from a function, where `n` is function's number of outputs according to its type defined in the type section.
+6. Maximum data stack height required by a function does not exceed `1024`.
+
+To examine every reachable code path, validation needs to traverse every instruction in order, while also following each non-conditional jump, and following both possible branches for each conditional jump. See below for reference implementation.
+
+The complexity of this traversal is linear in the number of instructions, because each code path is examined only once, and property 1 guarantees no loops in the validation.
+
+### Execution
+
+Given new deploy-time guarantees, EVM implementation is not required anymore to have run-time stack underflow check for each executed instruction.
+
+Stack overflow check, on the other hand, is still required at run-time, because function execution can start at arbitrary (i.e. known only at run-time) stack height at CALLF instruction of a caller (i.e. each execution can be in arbitrary inner call frame). Verification alrogirthm examines only stack height changes relative to starting stack height of the function.
+
+## Rationale
+
+TBA
+
+## Backwards Compatibility
+
+This change poses no risk to backwards compatibility, as it is introduced only for EOF1 contracts, for which deploying undefined instructions is not allowed, therefore there are no existing contracts using these instructions. The new instructions are not introduced for legacy bytecode (code which is not EOF formatted).
```

Comment:
Note: mention that this requires a hard fork

Reply:
Added.